extensions/azurecore/src/account-provider/auths/azureAuth.ts

```diff
@@ -0,0 +1,375 @@
+		try {
+			const tenantResponse = await this.makeGetRequest(token, tenantUri);
+			const tenants: Tenant[] = tenantResponse.data.value.map((tenantInfo: TenantResponse) => {
+				return {
+					id: tenantInfo.tenantId,
+					displayName: tenantInfo.displayName ? tenantInfo.displayName : localize('azureWorkAccountDisplayName', "Work or school account"),
+					userId: token.key,
+					tenantCategory: tenantInfo.tenantCategory
+				} as Tenant;
+			});
+
+			const homeTenantIndex = tenants.findIndex(tenant => tenant.tenantCategory === 'Home');
+			if (homeTenantIndex >= 0) {
+				const homeTenant = tenants.splice(homeTenantIndex, 1);
+				tenants.unshift(homeTenant[0]);
+			}
+
+			return tenants;
+		} catch (ex) {
+			console.log(ex);
+			throw new Error('Error retreiving tenant information');
+		}
+	}
+
+	protected async getToken(postData: { [key: string]: string }, scope: string, resource: Resource): Promise<TokenRefreshResponse | undefined> {
+		try {
+			const tokenUrl = `${this.loginEndpointUrl}${this.commonTenant}/oauth2/v2.0/token`;
+
+			const tokenResponse = await this.makePostRequest(tokenUrl, postData);
+			const tokenClaims = this.getTokenClaims(tokenResponse.data.access_token);
+
+			const accessToken: AccessToken = {
+				at: tokenResponse.data.access_token,
+				key: tokenClaims.email || tokenClaims.unique_name || tokenClaims.name,
+				resource: resource.id,
+			};
+
+			const refreshToken: RefreshToken = {
+				rt: tokenResponse.data.refresh_token,
+				key: accessToken.key,
+				resource: resource.id
+			};
+
+			return { accessToken, refreshToken, tokenClaims };
+
+		} catch (err) {
+			console.dir(err);
+			const msg = localize('azure.noToken', "Retrieving the token failed.");
+			throw new Error(msg);
+		}
+	}
+
+	protected getTokenClaims(accessToken: string): TokenClaims | undefined {
+		try {
+			const split = accessToken.split('.');
+			return JSON.parse(atob(split[1]));
+		} catch (ex) {
+			throw new Error('Unable to read token claims');
+		}
+	}
+
+	protected async refreshAccessTokens(account: azdata.Account, tokens: RefreshToken[]): Promise<void> {
+		for (let resource of this.resources) {
+			const properToken = tokens.find(t => t.resource === resource.id);
+			if (!properToken) {
+				account.isStale = true;
+				return;
+			}
+
+			this.refreshAccessToken(account.key, properToken, resource);
+		}
+	}
+
+	private async refreshAccessToken(account: azdata.AccountKey, token: RefreshToken, resource: Resource): Promise<void> {
+		const scopes = [...this.metadata.settings.scopes, resource.scopes];
+		const postData = {
+			grant_type: 'refresh_token',
+			refresh_token: token.rt,
+			client_id: this.clientId,
+			tenant: this.commonTenant,
+			scope: scopes.join(' ')
+		};
+
+		const { accessToken, refreshToken } = await this.getToken(postData, postData.scope, resource);
+
+		if (!accessToken || !refreshToken) {
+			console.log('Access or refresh token were undefined');
```

Comment:
will these errors be displayed to users or only in the dev console?

Reply:
Some of these errors are only dev console errors because they're not unexpected. Just makes debugging easier.